crates/pipeline_manager/src/prober/mod.rs

```diff
@@ -0,0 +1,118 @@
+use crate::api::{ManagerError, ServiceProbe, ServiceProbeResponse};
+use crate::config::ProberConfig;
+use crate::db::storage::Storage;
+use crate::db::{DBError, ProjectDB};
+use actix_web::{get, web, HttpServer, Responder};
+use log::{debug, error, info};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::Mutex;
+use tokio::time::sleep;
+use tokio::{join, spawn};
+
+/// Constantly checks for new probes being created in the database.
+/// It performs the probes sequentially one after the other (sequentially).
+/// If there are no probes to perform, it briefly sleeps before checking again.
+///
+/// Probes which are in `created` or `running` status are fetched one after the
+/// other ordered ascending by their `created_at` timestamp (thus, earliest
+/// created first). It also fetches `running` status, as currently there is only
+/// a single prober service, and it allows recovering if the prober service was
+/// unexpectedly terminated in the middle of a probe.
+///
+/// TODO: concurrent probing with a sliding window rather than sequential
+async fn service_probing_task(
+    config: ProberConfig,
+    db: Arc<Mutex<ProjectDB>>,
+) -> Result<(), ManagerError> {
+    info!(
+        "Probing service started (sleeps {} ms before checking again if there are no probes)",
+        config.probe_sleep_inbetween_ms
+    );
+    loop {
+        let result: Result<(), DBError> = {
+            loop {
+                let probe = db.lock().await.next_service_probe().await?;
```

Comment:
Any reason you're not using the listen_notify mechanism as the runners do? It would be simpler and more efficient than scanning the DB regularly.

Reply:
I modeled the implementation in the same fashion as the compiler, which also does the polling at interval. It looked like the simplest solution for the first version, and then improve / add the notify mechanism in a future PR.

Reply:
The compiler was written before we had the listen/notify mechanism, hence the change. Yes you can improve it in a subsequent PR.

Reply:
Added it to the list of future tasks in the main post of this PR.